### ISSUE TITLE
[SOLVED] Cleaning up tweens in a halted loop.

### ISSUE THREAD

**Issue by trusktr** [tweenjs/tween.js#241]
Suppose we have a requestAnimationFrame loop, and we're calling `TWEEN.update()` in there to update all tweens every tick. 

If we stop the loop (therefore stop calling `TWEEN.update()`), what happens to the tweens? What do you recommend for clean up? Or should we just leave the tweens as is, then if the loop is ever restarted, the timestamps may be past the duration of the tween and therefore the tweens clean themselves up?

And, on a related topic, how might we pause all tweens at once, then resume them later?

**Comment by mikebolt** [comment]
If you're worried about memory, you can use `TWEEN.removeAll()`. After that, nothing will happen on a `TWEEN.update()` until you add another tween.

> should we just leave the tweens as is, then if the loop is ever restarted, the timestamps may be past the duration of the tween and therefore the tweens clean themselves up?

Well, sort of. If you're updating with the timestamp argument in the requestAnimationFrame callback, then starting a new requestAnimationFrame loop will probably make all of the unfinished tweens jump to the end, or possibly just somewhere further along in the animation. If you have repeating tweens that would have finished by the time you start the new loop, then they will all start their next repetition at the same time. Also, note that tweens are never automatically removed, even when they finish. This may not be the intended behavior. I don't know. If you're worried about memory leaks you should put a call to `TWEEN.remove` in the `onComplete` callback.

If you want to pause everything, you will have to do it yourself. Basically, just stop calling TWEEN.update(), then when you call it again, pick up where you left off. See [my code from a pull request](https://github.com/mikebolt/tween.js/blob/master/examples/13_groups.html#L104) for an example of how to do this.

If you think that we should add a pause feature, then please make another issue for that. It should be easy to add that in.

**Comment by trusktr** [comment]
Yep, that's what I thought, custom timestamps for pausing/resuming. Cool, thanks! Suppose I make a tween, and let it animate by calling `TWEEN.update()` repeatedly until it's finished, and that tween is intended to be used only once.

So if I never ever call `TWEEN.remove` when tweens are finished, that'll leak memory? Maybe TWEEN could do that automatically? I'm imagining people might call a function that starts a tween, then return, forever losing the reference to the tween.

**Comment by mikebolt** [comment]
Really what happens is that tweens are just retained in memory until you remove them. So if you have some code that creates tweens in response to user input or some kind of event, then you need to ensure that they get removed or you will have a memory leak. But if you just have N tweens that you use once, then it might not be an issue (except the user's computer will have to keep holding on to those tweens until the browser window or tab is closed).

In light of this they should probably be automatically removed. If you keep a reference to the tween yourself, then you can restart it by just calling `start()` again, and there's no way to restart a tween without keeping a reference to it anyway (I'm not counting repetitions as restarting). This means that there really isn't any reason for the tweens to NOT be automatically removed. So I would say that this is either a bug or a design flaw.

**Comment by sole** [comment]
How can TWEEN know that you just stopped calling its update method "for a while" but intend to resume later? Likewise, how can it deduct that you actually are not going to call it ever again and the tweens can be removed? Should we set a timer and check if you haven't called update in a while and then remove tweens? But what if you legitimately wanted to pause updating for a while (e.g. maybe your tab went to the background)?

That's why things are not happening automatically---there's not a unique use case for everyone.

**Comment by mikebolt** [comment]
We were talking about two different things in this thread:

1. How to simulate pausing when using requestAnimationFrame

2. Why tweens are not removed after they finish doing everything expected of them

We don't have to predict anything. We don't know if the user "pauses" anything, and even if they did pause it we should keep the tweens in case the animation is resumed. I argue that we should remove tweens after they are definitely, provably finished and at which point it would be useless to retain them in memory.

**Comment by trusktr** [comment]
> Why tweens are not removed after they finish doing everything expected of them

Yep, I think in that case, TWEEN will know that the a tween is finished if it reached the end and was expected to fire the `onComplete` callback. But, as you mentioned @sole, if a tween isn't finished yet and calls to `update` stop, then TWEEN will simply consider the tween not to be finished (even if that pause is indefinite) so won't remove it.

**Comment by sole** [comment]
Tweens *should* be removed when they are done: https://github.com/tweenjs/tween.js/blob/e19a4c929bace618cd4a8ab1f36357f97a2353ae/src/Tween.js#L82

I'm sorry I'm a bit lost as to what we're discussing by now, but if the `TWEEN` object is retaining references to tween instances, then the memory consumption will grow over time, specially if you keep creating tweens (e.g. in a game loop).

If you find such a case and can provide a test demonstrating it, I'll appreciate it :-)

**Comment by trusktr** [comment]
Oh, for sure! Well, in that case, what *are* we talking about. :)
